Repository maintenance: CI and testing improvements

xBlaz3kx opened the pull request:
## Proposed changes

- Added local docker compose for unit and integration testing 
- Improved the CI pipeline - now has jobs for each test type

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before
merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/xBlaz3kx/ocppManager-go/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the CLA
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

## Further comments

lorenzodonini commented on docker-compose.test.yaml:
```diff
@@ -0,0 +1,40 @@
+      - TOXIPROXY_HOST=toxiproxy
+      - TOXIPROXY_PORT=8474
+      - PROXY_OCPP_LISTENER=
+      - PROXY_OCPP_UPSTREAM=
```
```suggestion
      - PROXY_OCPP_UPSTREAM=integration_test:8887
```

lorenzodonini commented on docker-compose.test.yaml:
```diff
@@ -0,0 +1,40 @@
+    environment:
+      - TOXIPROXY_HOST=toxiproxy
+      - TOXIPROXY_PORT=8474
+      - PROXY_OCPP_LISTENER=
```
```suggestion
      - PROXY_OCPP_LISTENER=toxiproxy:8886
```

lorenzodonini commented on Makefile:
```diff
@@ -0,0 +1,2 @@
+test:
+	docker compose -f docker-compose.test.yaml up -d
```
Instead of running a busy wait on the CI job, we could do this instead:
```
docker compose -f docker-compose.test.yaml up toxiproxy integration_test --abort-on-container-exit
```
this will also just run integration tests. WDYT?

lorenzodonini reviewed (changes requested):
Local and dockerized setup work differently (different networks), so we need to parametrize the test instance and toxiproxy with env variables accordingly. I put this code together that would need to be added to the `network_test.go` file to work for both environments:

```go
type proxyConfig struct {
	ToxiProxyHost     string `env:"TOXIPROXY_HOST" envDefault:"localhost"`
	ToxiProxyPort     int    `env:"TOXIPROXY_PORT" envDefault:"8474"`
	ProxyOcppListener string `env:"PROXY_OCPP_LISTENER" envDefault:"localhost:8886"`
	ProxyOcppUpstream string `env:"PROXY_OCPP_UPSTREAM" envDefault:"localhost:8887"`
}
```

and:

```go
func (s *NetworkTestSuite) SetupSuite() {
	var cfg proxyConfig
	err := env.Parse(&cfg)
	if err != nil {
		panic(err)
	}
	endpoint := fmt.Sprintf("%v:%v", cfg.ToxiProxyHost, cfg.ToxiProxyPort)
	client := toxiproxy.NewClient(endpoint)
        // .....
        p, err := client.CreateProxy("ocpp", cfg.ProxyOcppListener, cfg.ProxyOcppUpstream)
	require.NoError(s.T(), err)
	s.proxy = p
}
```

This needs a new support lib:
```
go get github.com/caarlos0/env/v11
```

lorenzodonini commented:
Thanks for the PR. If it's too many change requests to handle, do you want me to take the branch over?

xBlaz3kx commented:
> @xBlaz3kx Thanks for the PR. If it's too many change requests to handle, do you want me to take the branch over?

No worries, I will take some time and apply suggested changes.

xBlaz3kx commented:
I fixed the tests and also improved the testing workflow a bit: 
- Removed unnecessary waiting and go installation, since the teste are run via Docker
- Added SHA as postfix to artifacts to uniquely identify coverage files for each workflow run

Not sure why coveralls is failing though.

lorenzodonini commented on docker-compose.test.yaml:
```diff
@@ -0,0 +1,44 @@
+        go test -v -covermode=count -coverprofile=ocppj.out ./ocppj
+        go test -v -covermode=count -coverprofile=ocpp16.out -coverpkg=github.com/lorenzodonini/ocpp-go/ocpp1.6/... github.com/lorenzodonini/ocpp-go/ocpp1.6_test
+        go test -v -covermode=count -coverprofile=ocpp201.out -coverpkg=github.com/lorenzodonini/ocpp-go/ocpp2.0.1/... github.com/lorenzodonini/ocpp-go/ocpp2.0.1_test
+        sed '1d;$d' ocppj.out >> coverage.out
```
Coveralls upload is failing because of this, as it skips the `mode: count` line (required).

Easiest fix:
```suggestion
        sed '0d;$d' ocppj.out >> coverage.out
```

lorenzodonini commented on .github/workflows/test.yaml:
```diff
@@ -0,0 +1,59 @@
+          name: unit_coverage_${{ github.sha }}
+      - name: Run integration tests
+        run: |
+          docker compose -f docker-compose.test.yaml up integration_test --abort-on-container-failure
```
nitpick: the command never cleans up the toxiproxy container (which is started in the background and keeps running). By changing this to:
```
docker compose -f docker-compose.test.yaml up toxiproxy integration_test --abort-on-container-exit
```
all other containers will be stopped once the first container exists, which is imo cleaner. Note that `abort-on-container-exit` is key here, as `abort-on-container-failure` would just hang.

Same reasoning for the `unit` job.
